src/transformation/transformation.ts

```diff
@@ -3,9 +3,305 @@ import {
     ensureNodeDefinition,
     getNodeDefinition,
     Node,
-    NODE_DEFINITION_SYMBOL, registerNodeDefinition,
+    NODE_DEFINITION_SYMBOL, Origin, Property,
+    registerNodeDefinition,
     registerNodeProperty
 } from "../model/model";
+import {Issue, IssueSeverity} from "../validation";
+import {Position} from "../model/position";
+
+function isClassType(type) : boolean {
+    return typeof type == 'function' &&
+        type.name != undefined &&
+        new RegExp(`[class\\s+${type.name}.*]`).test(type.toString());
+}
+
+function isSuperClass(subClass, superClass) : boolean {
+    return isClassType(superClass) &&
+        // eslint-disable-next-line no-prototype-builtins
+        superClass.prototype.constructor.isPrototypeOf(subClass);
+}
+
+function getClassTypeName(type) : string | undefined {
+    if (isClassType(type))
+        return type.name;
+    else return undefined;
+}
+
+export class NodeFactory<Source, Output extends Node> {
+    constructor(
+        public constr: (s: Source, t: ASTTransformer, f: NodeFactory<Source, Output>) => Output | undefined,
+        public children: Map<string, ChildNodeFactory<Source, any, any> | undefined> = new Map(),
+        public finalizer: (Output) => void = () => undefined
+    ) {}
+
+    // TODO: port other overrides of the withChild method?
+    withChild = function<Target extends any, Child extends any>(
+       get: (s: Source) => any | undefined,
+       set: (t: Target, c?: Child) => void,
+       name: string,
+       type?: any
+    ) : NodeFactory<Source, Output> {
+
+        const prefix = isClassType(type) ? `${getClassTypeName(type)}#` : "";
+        this.children.set(prefix + name, new ChildNodeFactory(prefix + name, get, set));
+        return this;
+    }
+
+    withFinalizer = function (finalizer: (Output) => void) {
+        this.finalizer = finalizer;
+    }
+
+    getter : (Source) => any = function(path: string) {
+        return function(src: Source) {
+            let sub = src;
+
+            for (const elem in path.split(".")) {
+                if (sub == null)
+                    break;
+                sub = this.getSubExpression(sub, elem);
+            }
+
+            return sub;
+        }
+    }
+
+    private getSubExpression : any | undefined = function (src: any, elem: string) {
+        if (Array.isArray(src)) {
+            return src.map(it => this.getSubExpression(it!, elem));
+        } else {
+            const sourcePropName : string | undefined = Object.keys(src).find(e => e == elem);
+
+            if (!sourcePropName)
+                throw new Error(`${elem} not found in ${src} (class: ${Object.getPrototypeOf(src).constructor.name})`)
```

Comment:
I'm not sure this is necessary & I think we should explicitly test for objects with `undefined` properties.
The check exists in Kolasu because you need a property (meta)object to get/set values dynamically on the JVM, but here in TS/JS we don't need that.